### PR TITLE
My places tracks redesign

### PR DESCRIPTION
**Pull request by nnngrach**
[video](https://www.dropbox.com/scl/fi/gns78zs3bkoyue78l5ktc/2024-01-30-02.01.01.mov?rlkey=mbdh41md36jmi44hg9a7fye0a&dl=0)

Skipped:
Sorting
Navbar -> Select
Folder -> Details

### CODE REVIEW

File: Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//
+//  Created by Max Kojin on 02/02/24.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+class OATwoButtonsTableViewCell: OASimpleTableViewCell {
+    
+    @IBOutlet weak var leftButton: UIButton!
```
Comment (tigrim) [inline Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift:13]: private weak var (SwiftLint style)

File: Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @IBOutlet weak var rightButton: UIButton!
+    
+    func buttonsVisibility(show: Bool) {
+        leftButton.isHidden = !show
+        rightButton.isHidden = !show
+        updateMargins()
+    }
+    
+    func checkSubviewsToUpdateMargins() -> Bool {
+        return !leftButton.isHidden && !rightButton.isHidden
```
Comment (tigrim) [inline Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift:23]: remove return

File: Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    @IBOutlet weak var leftButton: UIButton!
+    @IBOutlet weak var rightButton: UIButton!
+    
+    func buttonsVisibility(show: Bool) {
+        leftButton.isHidden = !show
+        rightButton.isHidden = !show
+        updateMargins()
+    }
+    
+    func checkSubviewsToUpdateMargins() -> Bool {
```
Comment (tigrim) [inline Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift:22]: shouldUpdateMarginsForVisibleButtons?

File: Sources/Helpers/OAUtilities.m [hunk 4]
```diff
@@ -945,9 +945,13 @@ - (void) addBlurEffect:(BOOL)light cornerRadius:(CGFloat)cornerRadius padding:(C
 
 @implementation UITableViewCell (util)
 
-+ (NSString *) getCellIdentifier
++ (nonnull NSString *) getCellIdentifier
 {
-    return NSStringFromClass(self.class);
+    NSString *name = NSStringFromClass(self.class);
+    NSString *swiftClassPrefix = @"OsmAnd_Maps.";
```
Comment (tigrim) [inline Sources/Helpers/OAUtilities.m:951]: ?

File: Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+//  TwoButtonsTableViewCell.swift
+//  OsmAnd Maps
+//
+//  Created by Max Kojin on 02/02/24.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+class OATwoButtonsTableViewCell: OASimpleTableViewCell {
```
Comment (tigrim) [inline Sources/Controllers/Cells/OATwoButtonsTableViewCell.swift:11]: final

File: OsmAnd.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -16677,6 +16671,23 @@
...
+			);
+			inputPaths = (
+			);
+			outputFileListPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "# Type a script or drag a script file from your workspace to insert its path.\nkillall Simulator || echo \"Simulator app already closed\"\n";
```
Comment (tigrim) [inline OsmAnd.xcodeproj/project.pbxproj]: ?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 7]
```diff
@@ -122,13 +113,16 @@ class TracksViewController: OACompoundViewController, UITableViewDelegate, UITab
...
+        trackRecordingObserver = OAAutoObserverProxy.init(self, withHandler: #selector(onObservedRecordedTrackChanged), andObserve: app.trackRecordingObservable)
+        trackStartStopObserver = OAAutoObserverProxy.init(self, withHandler: #selector(onObservedRecordedTrackChanged), andObserve: app.trackStartStopRecObservable)
     }
     
     private func updateData() {
-        DispatchQueue.main.async { [weak self] in
-            self?.generateData()
-            self?.tableView.reloadData()
-        }
+//        DispatchQueue.main.async { [weak self] in
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:131]: ?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 8]
```diff
@@ -277,6 +271,16 @@ class TracksViewController: OACompoundViewController, UITableViewDelegate, UITab
         }
     }
     
+    private func setupNavbar() {
+        let title = folderName.length > 0 ? folderName : localizedString("menu_my_trips")
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:284]: folderName.isEmpty ?

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 9]
```diff
@@ -125,13 +125,16 @@ class TracksViewController: OACompoundViewController, UITableViewDelegate, UITab
...
+        tableView.register(UINib(nibName: OARightIconTableViewCell.reuseIdentifier, bundle: nil), forCellReuseIdentifier: OARightIconTableViewCell.reuseIdentifier)
+        tableView.register(UINib(nibName: OALargeImageTitleDescrTableViewCell.reuseIdentifier, bundle: nil), forCellReuseIdentifier: OALargeImageTitleDescrTableViewCell.reuseIdentifier)
     }
     
     private func updateData() {
-//        DispatchQueue.main.async { [weak self] in
-            self.generateData()
-            self.tableView.reloadData()
-//        }
+        self.generateData()
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:136]: remove self

File: Sources/Controllers/MyPlaces/TracksViewController.swift [hunk 10]
```diff
@@ -125,13 +125,16 @@ class TracksViewController: OACompoundViewController, UITableViewDelegate, UITab
...
+        self.tableView.reloadData()
```
Comment (tigrim) [inline Sources/Controllers/MyPlaces/TracksViewController.swift:137]: self

File: Sources/Helpers/OAGPXImportHelper.mm [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+- (void)doPush
+{
+    if (_hostVC)
+    {
+        parentController = _hostVC.parentViewController;
+        
+        CATransition* transition = [CATransition animation];
+        transition.duration = 0.4;
+        transition.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
+        transition.type = kCATransitionPush; // kCATransitionMoveIn; //, kCATransitionPush, kCATransitionReveal, kCATransitionFade
```
Comment (tigrim) [inline Sources/Helpers/OAGPXImportHelper.mm:119]: clean comments

File: Sources/Helpers/OAGPXImportHelper.mm [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+        transition.subtype = kCATransitionFromRight; //kCATransitionFromLeft, kCATransitionFromRight, kCATransitionFromTop, kCATransitionFromBottom
```
Comment (tigrim) [inline Sources/Helpers/OAGPXImportHelper.mm:120]: clean comments

File: Sources/Helpers/OAGPXImportHelper.mm [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+        [[OARootViewController instance].navigationController popToRootViewControllerAnimated:NO];
+    }
+}
+
++ (void)doPop
+{
+    CATransition* transition = [CATransition animation];
+    transition.duration = 0.4;
+    transition.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
+    transition.type = kCATransitionReveal; // kCATransitionMoveIn; //, kCATransitionPush, kCATransitionReveal, kCATransitionFade
```
Comment (tigrim) [inline Sources/Helpers/OAGPXImportHelper.mm:131]: clean comments

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/OASavingTrackHelper.h [hunk 14]
```diff
@@ -12,9 +12,17 @@
 #define kTrackNoHeading -1.0
 
 @class OAGPXMutableDocument;
+@class OAGPXDocument;
 @class OAGPX;
 @class OAWptPt;
 
+@protocol OATrackSavingHelperUpdatableDelegate <NSObject>
```
Comment (alex-osm) [inline Sources/Helpers/OASavingTrackHelper.h:19]: 1 - `OASavingTrackHelper` is backend class and should not have any UI code.
2 - It is not clear why added method are needed. They should be already present in current codebase. (Looks like need to be discussed)

File: Sources/Helpers/OAUtilities.m [hunk 15]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
+        [url getResourceValue:&isDirectory forKey:NSURLIsDirectoryKey error:nil];
+        if (isDirectory && [isDirectory boolValue])
+            return YES;
     }
+    return NO;
+}
+
++ (BOOL) isDirByPath:(NSString *)path
+{
+    NSURL *url = [NSURL fileURLWithPath:path];
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2656]: `(BOOL)fileExistsAtPath:(NSString *)path isDirectory:(nullable BOOL *)isDirectory;` ?

File: Sources/Helpers/OAUtilities.m [hunk 16]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
+                NSMutableArray<NSString *> *recursiveFoundSubfolderPathes = [self.class getSubfoldersList:subfolderPath];
+                if (recursiveFoundSubfolderPathes.count > 0)
+                    [allSubfolderPathes addObjectsFromArray:recursiveFoundSubfolderPathes];
+            }
+        }
+    }
+    return allSubfolderPathes;
+}
+
++ (BOOL) isDirByUrl:(NSURL *)url
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2642]: Is it needed?

File: Sources/Helpers/OAUtilities.m [hunk 17]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
+        {
+            if([subfolderUrl.lastPathComponent hasPrefix:@"."])
+                continue;
+            
+            if ([self.class isDirByUrl:subfolderUrl])
+            {
+                NSString *subfolderPath = [path stringByAppendingPathComponent:subfolderUrl.lastPathComponent];
+                [allSubfolderPathes addObject:subfolderPath];
+                
+                NSMutableArray<NSString *> *recursiveFoundSubfolderPathes = [self.class getSubfoldersList:subfolderPath];
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2633]: 1 - Do no use word `recursive`
2 - Paths

File: Sources/Helpers/OAUtilities.m [hunk 18]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
         }];
     }
     else
     {
-        return [NSArray arrayWithArray:allFoldersNames];
+        return [NSArray arrayWithArray:allFoldersShortPathes];
+    }
+}
+
++ (NSMutableArray<NSString *> *) getSubfoldersList:(NSString *)path
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2611]: `getFlattenedFileList`

File: Sources/Helpers/OAUtilities.m [hunk 19]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
     return [trackFolderName stringByAppendingPathComponent:fullFilePath.lastPathComponent];
 }
 
-+ (NSArray<NSString *> *) getGpxFoldersListSorted:(BOOL)shouldSort shouldAddTracksFolder:(BOOL)shouldAddTracksFolder
++ (NSArray<NSString *> *) getGpxFoldersListSorted:(BOOL)shouldSort shouldAddRootTracksFolder:(BOOL)shouldAddRootTracksFolder
 {
-    NSMutableArray<NSString *> *allFoldersNames = [NSMutableArray new];
-    NSArray* filesList = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:OsmAndApp.instance.gpxPath error:nil];
-    for (NSString *name in filesList)
+    NSMutableArray<NSString *> *allFoldersFullPathes = [self.class getSubfoldersList:OsmAndApp.instance.gpxPath];
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2583]: `allFoldersFullPathes` > `flattenedFilePaths`

File: Sources/Helpers/OAUtilities.m [hunk 20]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
+    NSMutableArray<NSString *> *allFoldersShortPathes = [NSMutableArray array];
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2584]: `flattenedRelativeFilePaths`

File: Sources/Helpers/OAUtilities.m [hunk 21]
```diff
@@ -2578,28 +2578,83 @@ + (NSString *) getGpxShortPath:(NSString *)fullFilePath
...
+    NSMutableArray<NSString *> *allSubfolderPathes = [NSMutableArray new];
+    
+    NSString *currentFolderPath = path;
+    if (!currentFolderPath || currentFolderPath.length == 0)
+        currentFolderPath = OsmAndApp.instance.documentsPath;
+    
+    NSURL *currentFolderUrl = [NSURL fileURLWithPath:currentFolderPath];
+    if (currentFolderUrl)
+    {
+        NSArray<NSURL *> *subfolderUrls =  [[NSFileManager defaultManager] contentsOfDirectoryAtURL:currentFolderUrl includingPropertiesForKeys:@[NSURLNameKey, NSURLIsDirectoryKey] options:NSDirectoryEnumerationSkipsHiddenFiles error:nil];
```
Comment (alex-osm) [inline Sources/Helpers/OAUtilities.m:2622]: Can it be replaced with `subpathsOfDirectoryAtPath`?

**Pull request merged**